backend/src/worker.rs

```diff
@@ -530,138 +551,514 @@ async fn handle_nondep_job(
 
-                let child = if !disable_nsjail {
-                    Command::new(nsjail_path)
-                        .current_dir(job_dir)
-                        .args(vec!["--config", "download.config.proto"])
-                        .stdout(Stdio::piped())
-                        .stderr(Stdio::piped())
-                        .spawn()?
-                } else {
-                    Command::new(python_path)
-                        .current_dir(job_dir)
-                        .args(vec![
-                            "-m",
-                            "pip",
-                            "install",
-                            "--no-color",
-                            "--isolated",
-                            "--no-warn-conflicts",
-                            "--disable-pip-version-check",
-                            "-t",
-                            "./dependencies",
-                            "-r",
-                            "./requirements.txt",
-                        ])
-                        .stdout(Stdio::piped())
-                        .stderr(Stdio::piped())
-                        .spawn()?
-                };
+        install_go_dependencies(
+            &job.id,
+            inner_content,
+            logs,
+            job_dir,
+            status,
+            db,
+            last_line,
+            timeout,
+            go_path,
+            true,
+        )
+        .await?;
+    }
 
-                logs.push_str("\n--- PIP DEPENDENCIES INSTALL ---\n");
-                *status = handle_child(job, db, logs, last_line, timeout, child).await;
-                tracing::info!(
-                    worker_name = %worker_name,
-                    job_id = %job.id,
-                    workspace_id = %job.workspace_id,
-                    is_ok = status.is_ok(),
-                    "finished setting up python dependencies {}",
-                    job.id
-                );
-            }
-            if requirements.len() == 0 || status.is_ok() {
-                logs.push_str("\n\n--- PYTHON CODE EXECUTION ---\n");
+    logs.push_str("\n\n--- GO CODE EXECUTION ---\n");
+    set_logs(logs, job.id, db).await;
+
+    let sig = crate::parser_go::parse_go_sig(&inner_content)?;
+    let token = create_token_for_owner(
+        &db,
+        &job.workspace_id,
+        &job.permissioned_as,
+        "ephemeral-script",
+        timeout * 2,
+        &job.created_by,
+    )
+    .await?;
+    let args = if let Some(args) = &job.args {
+        Some(
+            transform_json_value(&token, &job.workspace_id, &base_internal_url, args.clone())
+                .await?,
+        )
+    } else {
+        None
+    };
+    let ser_args = serde_json::to_string(&args).map_err(|e| Error::ExecutionErr(e.to_string()))?;
+    write_file(job_dir, "args.json", &ser_args).await?;
+    let spread = sig
+        .args
+        .into_iter()
+        .map(|x| {
+            format!(
+                "json_arg[\"{}\"].({})",
```

Comment:
String interpolation here freaks me out.

I suggest doing the spread in rust by the worker.

Supposing that args is a `Map<String, Value>`, I suspect we can just put it in order as a `Vec<Value>` parallel with `sig.args` and write that array to `args.json`. In the Go program, we unmarshal into `[interface{}]` or something and use [reflect.Call](https://pkg.go.dev/reflect@go1.19.1#Value.Call) to call the user function?

Reply:
Let's sync about this, I am not sure to understand fully:
- What is wrong with string interpolation here
- What is the suggested solution

Reply:
Here is a simpler suggestion that doesn't use reflect and still uses spread but does not interpolate user input.

```rust
    let args = if let Some(args) = &job.args {
        let args =
            transform_json_value(&token, &job.workspace_id, &base_internal_url, args.clone())
                .await?;
        sig.args.iter().map(|a| args[&a.name].clone()).collect()
    } else {
        Vec::default()
    };

    let spread = (0..args.len()).map(|i| format!("json_arg[{i}]")).join(",");

    let ser_args = serde_json::to_string(&Value::Array(args))
        .map_err(|e| Error::ExecutionErr(e.to_string()))?;
```

a more complete diff

```rust
diff --git a/backend/src/worker.rs b/backend/src/worker.rs
index f6d99ed..b7b2898 100644
--- a/backend/src/worker.rs
+++ b/backend/src/worker.rs
@@ -690,27 +690,21 @@ async fn handle_go_job(
         &job.created_by,
     )
     .await?;
+
     let args = if let Some(args) = &job.args {
-        Some(
+        let args =
             transform_json_value(&token, &job.workspace_id, &base_internal_url, args.clone())
-                .await?,
-        )
+                .await?;
+        sig.args.iter().map(|a| args[&a.name].clone()).collect()
     } else {
-        None
+        Vec::default()
     };
-    let ser_args = serde_json::to_string(&args).map_err(|e| Error::ExecutionErr(e.to_string()))?;
+
+    let spread = (0..args.len()).map(|i| format!("json_arg[{i}]")).join(",");
+
+    let ser_args = serde_json::to_string(&Value::Array(args))
+        .map_err(|e| Error::ExecutionErr(e.to_string()))?;
     write_file(job_dir, "args.json", &ser_args).await?;
-    let spread = sig
-        .args
-        .into_iter()
-        .map(|x| {
-            format!(
-                "json_arg[\"{}\"].({})",
-                x.name,
-                x.otyp.unwrap_or_else(|| "interface{}".to_string())
-            )
-        })
-        .join(", ");

     let inner_content = inner_content.replace("func main(", "func inner_main(");
     let wrapper_content: String = format!(
@@ -731,7 +725,7 @@ func main() {{
         os.Exit(1)
     }}

-    var json_arg map[string]interface{{}}
+    var json_arg []interface{{}}

     if err := json.Unmarshal(dat, &json_arg); err != nil {{
         fmt.Println(err)
```

Reply:
Reflect could be useful to support multiple return types but has the disadvantage of moving compile errors to runtime and it's probably out of scope.

Reply:
You're right, type assertions are needed on the arguments. So something like this works

```rust
    let spread = sig
        .args
        .iter()
        .enumerate()
        .map(|(i, arg)| {
            let assertion = arg
                .otyp
                .as_ref()
                .map(|ty| format!(".({ty})"))
                .unwrap_or_default();
            format!("json_arg[{i}]{assertion}")
        })
        .join(",");
```

So this removes the argument identifiers from the generated code but is otherwise the same.

Reply:
I think it's equivalent and not significantly better to use map directly rather than list